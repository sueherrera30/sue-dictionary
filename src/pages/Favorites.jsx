import React, { Fragment } from "react";
import { useNavigate } from 'react-router-dom';
import { MdOutlineArrowBack as Back } from "react-icons/md";
import { IconButton } from '@mui/material';
import ClearIcon from '@mui/icons-material/Clear';

import hearts from '../assets/favs.png';
import thinking from '../assets/thinking.png'
import sharedStyles from './styles/home.module.css';
import styles from './styles/favorites.module.css';
const Favorites = ({ favorites, setFavorites }) => {
    let navigate = useNavigate();

    const goToDefinition = (word) => {
        navigate(`/definition/${word}`)
    };

    const deleteWord = (word) => {
        const filteredFavs = favorites.filter(item => item !== word);
        setFavorites(filteredFavs);
    }

    return (
        <section className={sharedStyles.mainContainer}>
            <div className={`${sharedStyles.container} ${styles.container}`}>
                <header className={styles.header}>
                    <Back data-testid="back-button" className={styles.icons} onClick={() => navigate('/')} />
                    <div className={styles.imageContainer} >
                        <img alt="decoration" src={thinking} className={styles.image} />
                        <img src={hearts} alt="hearts" className={styles.hearts}/>
                    </div>
                    <h2 className={styles.title}>MY FAVORITES WORDS</h2>
                </header>
                <div className={styles.wordsContainer} >
                    {
                        favorites.length !== 0 ?
                        favorites.map((favWord, idx) => (
                            <Fragment key={`${favWord}-${idx}`}>
                                <div data-testid="word-saved" onClick={() => goToDefinition(favWord)} className={styles.word}>
                                    {favWord}
                                </div>
                                 <IconButton data-testid="delete-button" onClick={() => deleteWord(favWord)}><ClearIcon className={styles.DeleteIcon} /></IconButton>
                            </Fragment>
                        )) : <p  className={styles.message}>there are no favorites words saved, go ahead!</p>
                    }
                </div>
            </div>
        </section>
    )
};

export default Favorites;
