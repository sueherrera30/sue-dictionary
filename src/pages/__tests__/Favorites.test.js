import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import * as router from 'react-router'
import Favorites from '../Favorites';
import userEvent from '@testing-library/user-event'

const fovarites = ['test', 'word','love'];
const setFavorites = jest.fn();
const ui = userEvent.setup()
const navigate = jest.fn()

// const mockSetState = jest.fn();
// jest.mock('react', () => ({
//   useState: initial => [initial, mockSetState]
// }));

const deleteWord = jest.fn(() => Favorites);


beforeEach(() => {
  jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
})
afterEach(() => {
    cleanup();
});
test('<Favorites />', () => {
    render(<MemoryRouter><Favorites favorites={fovarites} setFavorites={setFavorites} /></MemoryRouter>, );
    expect(screen.container).toMatchSnapshot();
    expect(screen.getByText(/MY FAVORITES WORDS/i)).toBeInTheDocument();
});

test('Shows list of words saved', () => {
    render(<MemoryRouter><Favorites favorites={fovarites} /></MemoryRouter>,);
    const words = screen.queryAllByTestId('word-saved');
    expect(words.length).toBe(fovarites.length);
});

test('Goes to definition page', async () => {
  render(<MemoryRouter><Favorites favorites={fovarites} /></MemoryRouter>,);
  const word = screen.queryAllByTestId('word-saved')[0].textContent;
  await ui.click(screen.queryAllByTestId('word-saved')[0])
  expect(navigate).toHaveBeenCalledWith(`/definition/${word}`)
});

test('Back to home', async () => {
  render(<MemoryRouter><Favorites favorites={fovarites} /></MemoryRouter>,);
  const backButton = screen.getByTestId('back-button');
  await fireEvent.click(backButton);
  expect(navigate).toHaveBeenCalledWith('/')
});

test('Delete favorite word', async () => {
  render(<MemoryRouter><Favorites favorites={fovarites} setFavorites={deleteWord} /></MemoryRouter>,);
  const deleteButton = screen.queryAllByTestId('delete-button')[0];
  await fireEvent.click(deleteButton);
  expect(screen.queryAllByTestId('word-saved')[0].textContent).toBe('test');
});