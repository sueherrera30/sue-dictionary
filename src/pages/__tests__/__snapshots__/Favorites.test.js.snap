// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorites /> 1`] = `undefined`;
